### PR TITLE
RK2

### PR DESCRIPTION
**Pull request by almkuznetsov**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+/node_modules/
```
Comment (SergTyapkin) [inline .gitignore]: Почему не просто `node_modules` то?)

File: public/components/list/list.hbs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                                <div class="adElTopDecoration"></div>
+                                <div class="adElInnerBackground"></div>
+                            </div>
+                            <div class="adElDay">21</div>
+                            <div class="adElWeekday">Wed</div>
+                            <div class="adElTitle">Bento 3D</div>
+                            <div class="adElSubtitle">Ui Design</div>
+                        </div>
+
+                        <div class="adElContainer">
```
Comment (SergTyapkin) [inline public/components/list/list.hbs]: Все классы должны писаться в одном кейсе. Либо camelCase. Либо cebab-case.

File: public/components/list/list.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        this.SubmitCallback = submitCallback;
+        this.form = null;
+        this.errorLabel = null
+    }
+
+    render() {
+        const link = document.createElement('link');
+        link.rel = 'stylesheet';
+        link.type = 'text/css';
+        link.href = '../../static/css/list.css';
```
Comment (SergTyapkin) [inline public/components/list/list.js]: Когда добавите сборщик, тут будет попоболь

File: public/components/signup/signup.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import { Api } from "../../modules/api.js";
+import { Validate } from "../../modules/validate.js"
+
+export class Signup {
+    constructor(parent = document.body, submitCallback = () => {}) {
+        this.parent = parent;
+        this.SubmitCallback = submitCallback;
+        this.form = null;
+        this.errorLabel = null
```
Comment (SergTyapkin) [inline public/components/signup/signup.js]: ;

Comment (SergTyapkin) [reply public/components/signup/signup.js]: У вас же есть линтер - ну прогоните вы `eslint --fix`

File: public/components/signup/signup.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            }
+
+        });
+
+        Api.signup(inputsValue).then(
+            (response) => {
+                if (response.status < 300) {
+                    this.SubmitCallback();
+                } else {
+                    this.showError("Неверные данные")
```
Comment (SergTyapkin) [inline public/components/signup/signup.js]: eslint

File: public/components/signup/signup.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                }
+            }
+        );
+    }
+
+    showError(message){
+        this.errorLabel.style.visibility = "visible";
```
Comment (SergTyapkin) [inline public/components/signup/signup.js]: Через js ставить style - очень плохая практика. Ставим только классы

Comment (SergTyapkin) [reply public/components/signup/signup.js]: Через js можно ставить в `style` только css переменные. В случаях, когда у нас не два положения (вкл / выкл) а какой-нибудь изменяемый прогресс чего-то. И в `.css` прописать надо будет зависимости от этой переменной через `var(--some-variable)`

File: public/components/signup/signup.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        const link = document.createElement('link');
+        link.rel = 'stylesheet';
+        link.type = 'text/css';
+        link.href = '../../static/css/signup.css';
+        document.head.appendChild(link);
+        this.parent.innerHTML = Handlebars.templates["signup.hbs"]();
+        this.form = this.parent.getElementsByClassName("signup")[0];
+        this.form.addEventListener("submit", this.onSubmit.bind(this));
+        this.errorLabel = this.form.getElementsByClassName("error-label")[0];
+        this.errorLabel.style.visibility = "hidden";
```
Comment (SergTyapkin) [inline public/components/signup/signup.js]: См. ниже

File: public/index.html [hunk 8]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>AdHub</title>
+   <!-- <link rel="stylesheet" href="../../static/css/login.css"> -->
+    <!-- <link rel="stylesheet" href="../../static/css/signup.css"> -->
+    <link rel="preconnect" href="https://fonts.googleapis.com"> 
+<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
```
Comment (SergTyapkin) [inline public/index.html]: Что с отступами?

File: public/index.html [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    <!-- <link rel="stylesheet" href="../../static/css/signup.css"> -->
+    <link rel="preconnect" href="https://fonts.googleapis.com"> 
+<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin> 
+<link href="https://fonts.googleapis.com/css2?family=Istok+Web&display=swap" rel="stylesheet">
+</head>
+<body>
+<div id="root">
+
+</div>
+<img id="image" class="logo-right-bottom" class="list-none" src="static/img/logo_ad.png" alt="Описание изображения">
```
Comment (SergTyapkin) [inline public/index.html]: отступы

Comment (SergTyapkin) [reply public/index.html]: Поправить `alt="..."`

File: public/index.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import { Login } from "./components/login/login.js";
+import { Signup } from "./components/signup/signup.js";
+import { List } from "./components/list/list.js";
+
+
+let router = new Router();
+
+let login = new Login(document.getElementById("root"), () => router.go("/list"))
+let signup = new Signup(document.getElementById("root"), () => router.go("/login"))
+let list = new List(document.getElementById("root"), () => router.go("/index"))
```
Comment (SergTyapkin) [inline public/index.js]: Прогнать линтер --fix

Comment (SergTyapkin) [reply public/index.js]: Почему let, а не const

File: public/modules/ajax.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                    status,
+                    parsedJson,
+                };
+            });
+    }
+
+    static post(url = "", data = {}) {
+        let status;
+
+        return fetch(url, {
```
Comment (SergTyapkin) [inline public/modules/ajax.js]: Когда у вас нет `.catch` в обработке промиса, лучше использовать `async-await` синтаксис.

File: public/modules/ajax.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+            .then((parsedJson) => {
+                return {
+                    status,
+                    parsedJson,
+                };
+            });
+    }
+
+    static async postAsync(url = "", data = {}) {
+        const response = await fetch(url, {
```
Comment (SergTyapkin) [inline public/modules/ajax.js]: Ну типа да, как тут

File: public/modules/api.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+const URLS = {
+    login: "/auth",
+    signup: "/user",
+};
+
+const BACKEND_URL = "http://localhost:8080";
+
+export class Api{
+    static login(data={}){
```
Comment (SergTyapkin) [inline public/modules/api.js]: Линтер

File: public/modules/router.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
+export class Router {
+    constructor(routes = new Map) {
+        this.routes = routes;
+        this.current = null;
+    }
+
+    go(path) {
+        console.log("go  " + path)
+        if (this.current === path) return;
```
Comment (SergTyapkin) [inline public/modules/router.js]: `return` на новой строке. Линтер вроде тоже должен уметь такое редактировать

File: public/modules/router.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+            return;
+        }
+        this.routes.get(path)();
+    }
+
+    add(path, callback) {
+        this.routes.set(path, callback);
+    }
+
+    start() {
```
Comment (SergTyapkin) [inline public/modules/router.js]: В идеале вспоминаем Си, где если ты что-то создаешь, то ты должен это и очистить. В JS это справедливо в отношении eventListner'ов. Должен быть метод stop() с удалением listner'а

File: public/modules/validate.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        return emailPattern.test(email);
+    }
+
+    /**
+     * Валидация данных пароля
+     * @param {string} password - Введенные данные
+     */
+
+    static Password(password) {
+        return password.length > 5 && /^[a-zA-Z0-._-]+$/.test(password)
```
Comment (SergTyapkin) [inline public/modules/validate.js]: `return /^[a-zA-Z0-._-]{5,}$/.test(password)`

Comment (SergTyapkin) [reply public/modules/validate.js]: У вас могло не работать, если вы написали `{5, }` (с пробелом после запятой)

File: public/static/img/image_index.png
Comment (SergTyapkin) [inline public/static/img/image_index.png:1]: Лучше оставить одну картинку для фона, и поверх положить элемент с `background: linear-gradient()`. Тогда у вас и картинки не будут прыгать при переходах, и фон можно будет плавно заанимировать

**Review by SergTyapkin — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 17]
```diff
@@ -0,0 +1,6 @@
+node_modules
+bundle
+.idea
+.DS_Store
+public/main.bundle.js
+public/main.bundle.js.LICENSE.txt
```
Comment (SergTyapkin) [inline .gitignore:6]: Выпилить

File: package.json [hunk 18]
```diff
@@ -0,0 +1,106 @@
+{
+  "name": "adhub",
+  "version": "1.0.0",
+  "description": "PCF team project",
+  "main": "index.js",
+  "dependencies": {
```
Comment (SergTyapkin) [inline package.json:6]: Не верю, что их так много и все используются

File: package.json [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "start": "node server/index.js",
+    "build-templates": "handlebars public/components/login/login.hbs -f public/components/login/login.precompiled.js && handlebars public/components/signup/signup.hbs -f public/components/signup/signup.precompiled.js && handlebars public/components/list/list.hbs -f public/components/list/list.precompiled.js",
+    "eslint": "npx eslint .",
+    "dev": "webpack serve --mode development",
+    "build": "webpack --mode production"
+  },
+  "author": "PCF",
+  "license": "ISC",
+  "devDependencies": {
```
Comment (SergTyapkin) [inline package.json:87]: Всё кроме eslint в dependencies. Без них проект не соберется

File: public/components/editpage/editpage.js [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+    this.errorLabel = null;
+  }
+
+  render() {
+
+    Api.getAd(adID).then(
+    (data) => { 
+      console.log(data);
+      context.ad = data.parsedJson;
+    }
```
Comment (SergTyapkin) [inline public/components/editpage/editpage.js:28]: отступы

File: public/components/editpage/editpage.js [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+    });
+  
+    this.form.querySelector('#description').addEventListener('input', (event) => {
+      const previewDescription = document.querySelector('.editpage__preview-desription'); 
+      if (previewDescription) {
+        previewDescription.textContent = event.target.value; 
+      }
+    });
+
+    document.querySelector('#submitBtn').addEventListener('click', () => {
```
Comment (SergTyapkin) [inline public/components/editpage/editpage.js:103]: почему не getElementById?

File: public/components/editpage/editpage.js [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+
+  try {
+      const response = fetch(BACKEND_URL + "/adedit", requestOptions)        
+      .then(response => {
+        if (response.status < 300) {
+            location.href = "/company";
+        } else {
+            // Обработка ошибки
+            return response.json();
+        }})
```
Comment (SergTyapkin) [inline public/components/editpage/editpage.js:145]: Стиль кода - ужас просто, а так же отступы везде. ; тоже периодически нет

File: public/components/menu/menu.hbs [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+                    <p>Profile</p>
+                    <li>
+                        <div class="menu-element">
+                            <img src="../../static/img/free-icon-user-4743186.png" alt="Иконка">
+                            <a href="/profile">Профиль</a>
+                        </div>
+                    </li>
+                    <li>
+                        <div class="menu-element">
+                            <img src="../../static/img/free-icon-notification-8860191.png" alt="Иконка">
```
Comment (SergTyapkin) [inline public/components/menu/menu.hbs:50]: Такие названия файлов - просто не круто

File: public/components/menu/menu.js [hunk 24]
```diff
@@ -0,0 +1,13 @@
+import '../../static/css/menu.css'
+export default class Menu {
+    constructor(parent = document.body, submitCallback = () => {}) {
+        this.parent = parent;
+        this.SubmitCallback = submitCallback;
```
Comment (SergTyapkin) [inline public/components/menu/menu.js:5]: Почему с большой буквы?

File: public/index.html [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+    <link rel="preconnect" href="https://fonts.googleapis.com">
+    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
+    <link href="https://fonts.googleapis.com/css2?family=Istok+Web&display=swap" rel="stylesheet">
+</head>
+<body>
+<script src="/handlebars/dist/handlebars.runtime.js"></script>
+<div id="root">
+
+</div>
+<img id="image" class="logo-right-bottom" class="list-none" src="static/img/logo_ad.png" alt="Логотип">
```
Comment (SergTyapkin) [inline public/index.html:15]: отсутпы

File: public/index.js [hunk 26]
```diff
@@ -0,0 +1,48 @@
+'use strict';
+
+import Api from './modules/api.js';
+import './static/css/signup.css'
```
Comment (SergTyapkin) [inline public/index.js:4]: Этого здесь быть не должно

File: public/modules/ajax.js [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+      mode: 'cors',
+      credentials: 'include',
+    });
+    const parsedJson = await response.json();
+
+    return {
+      status: response.status,
+      parsedJson,
+    };
+  }
```
Comment (SergTyapkin) [inline public/modules/ajax.js:96]: Много дублирования кода. Лучше вынести в ещё одну функцию

File: public/modules/api.js [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+  static login(data={}) {
+    return Ajax.post(BACKEND_URL + URLS.login, data);
+  }
+
+  static signup(data={}) {
+    return Ajax.post(BACKEND_URL + URLS.signup, data);
+  }
+
+  static getAdsList() {
+    return Ajax.get({url:BACKEND_URL + '/ad'});
```
Comment (SergTyapkin) [inline public/modules/api.js:22]: пробел после :

File: public/modules/api.js [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+  signup: '/user',
+  ad: '/ad',
+  useredit: '/useredit',
+};
+
+export const BACKEND_URL = 'http://127.0.0.1:8080/api/v1';
+
+export default class Api {
+  static login(data={}) {
+    return Ajax.post(BACKEND_URL + URLS.login, data);
```
Comment (SergTyapkin) [inline public/modules/api.js:14]: У оригинального Ajax можно настроить BASE_URL_PATH изначально и не плюсовать его каждый раз. Сделайте так же

File: public/static/css/createad.css
Comment (SergTyapkin) [inline public/static/css/createad.css:1]: А почему css а не sass?

File: public/static/css/login.css [hunk 30]
```diff
@@ -0,0 +1,109 @@
+.login_body {
+    font-family: 'Montserrat', sans-serif;
```
Comment (SergTyapkin) [inline public/static/css/login.css:2]: Шрифты хотя бы точно вынесите в миксины. И через sass

File: public/static/css/my.css [hunk 31]
```diff
@@ -0,0 +1,19 @@
...
+html, body {
+    margin: 0;
+    padding: 0;
+}
+
+h1, h2, h3 {
+    font-family: 'Istok Web', sans-serif;
+    font-weight: 700; /* Применить полужирное начертание к заголовкам (если необходимо) */
+}
+  body {
```
Comment (SergTyapkin) [inline public/static/css/my.css:14]: Отступы

**Review by SergTyapkin — APPROVED** [review]
